### PR TITLE
ref: Refactor replace_group to class-based types

### PR DESCRIPTION
**Pull request by untitaker**
Also improve typing around _build_event_set_filter

Previously: https://github.com/getsentry/snuba/pull/1968

why now again? well, because I left the job half-finished and it really bugged me when I tried to add types from sentry-kafka-schemas

### CODE REVIEW

File: snuba/replacers/errors_replacer.py [hunk 1]
```diff
@@ -478,50 +475,103 @@ def get_timestamp_condition(msg_field: str, operator: str) -> str:
...
+    def get_replacement_type(self) -> ReplacementType:
+        return self.replacement_type
 
-    full_where = f"PREWHERE {' AND '.join(prewhere)} WHERE {' AND '.join(where)}"
-    project_id: int = message.data["project_id"]
-    query_time_flags = (None, project_id)
+    def get_message_metadata(self) -> ReplacementMessageMetadata:
+        return self.replacement_message_metadata
 
-    return _build_group_replacement(
```
Comment (untitaker) [inline snuba/replacers/errors_replacer.py:517]: _build_group_replacement was a bad abstraction that got in the way of refactoring, so I just inlined the relevant bits.

**Comment by lynnagara** [comment]
This seems a lot cleaner in general though all the Replacement vs LegacyReplacement stuff makes it confusing... is there any reason we can't just blow away the legacy now? It's been 2 years since it was introduced so it feels like time.

**Comment by untitaker** [comment]
> is there any reason we can't just blow away the legacy now?

that's the goal, but at the time it was decided that it shouldn't be one PR effectively rewriting this entire file, and that we should make incremental progress towards getting rid of function-based replacements. this is the reason LegacyReplacement was built as a compat layer.

**Comment by lynnagara** [comment]
I see... so this is a step on the way to removing LegacyReplacement?

File: snuba/replacers/errors_replacer.py [hunk 2]
```diff
@@ -478,50 +475,106 @@ def get_timestamp_condition(msg_field: str, operator: str) -> str:
 
     query_args = {
         "event_ids": event_id_list,
-        "project_id": message["project_id"],
+        "project_id": str(project_id),
```
Comment (lynnagara) [inline snuba/replacers/errors_replacer.py:478]: Is this changing to string? How come?

File: snuba/replacers/errors_replacer.py [hunk 3]
```diff
@@ -478,50 +475,106 @@ def get_timestamp_condition(msg_field: str, operator: str) -> str:
...
+            from_timestamp=self.from_timestamp,
+            to_timestamp=self.to_timestamp,
+        )
+
+        return (
+            f"PREWHERE {' AND '.join(prewhere)} WHERE {' AND '.join(where)}"
+            % query_args
+        )
+
+    def get_count_query(self, table_name: str) -> Optional[str]:
```
Comment (lynnagara) [inline snuba/replacers/errors_replacer.py:554]: Curious about this interface - how come `table_name` isn't just a property on the replacement

**Comment by untitaker** [comment]
> I see... so this is a step on the way to removing LegacyReplacement?

yeah the goal is to get rid of all uses of LegacyReplacement for sure

File: snuba/replacers/errors_replacer.py (hunk 3 above)
Comment (untitaker) [reply snuba/replacers/errors_replacer.py:554]: I did that just to satisfy typechecking, and I don't think it matters for runtime. those parameters are string-interpolated

File: snuba/replacers/errors_replacer.py [hunk 4]
```diff
@@ -478,50 +475,106 @@ def get_timestamp_condition(msg_field: str, operator: str) -> str:
...
+            SELECT count()
+            FROM {table_name} FINAL
+            {self._where_clause}
+        """
+
+    def get_insert_query(self, table_name: str) -> Optional[str]:
+        all_column_names = [c.escaped for c in self.all_columns]
+        select_columns = ", ".join(
+            map(
+                lambda i: i if i != "group_id" else str(self.new_group_id),
```
Comment (lynnagara) [inline snuba/replacers/errors_replacer.py:565]: Similar question to the project id, how come it's `str`?

**Review by lynnagara — APPROVED** [review]

File: snuba/replacers/errors_replacer.py (hunk 4 above)
Comment (untitaker) [reply snuba/replacers/errors_replacer.py:565]: tbh i think this was already like this?

**Pull request merged**